java/ql/src/Security/CWE/CWE-798/HardcodedCredentialsApiCall.ql

```diff
@@ -26,9 +27,13 @@ class HardcodedCredentialApiCallConfiguration extends DataFlow::Configuration {
 
   override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
     node1.asExpr().getType() instanceof TypeString and
-    exists(MethodAccess ma | ma.getMethod().hasName(["getBytes", "toCharArray"]) |
-      node2.asExpr() = ma and
-      ma.getQualifier() = node1.asExpr()
+    (
+      exists(MethodAccess ma | ma.getMethod().hasName(["getBytes", "toCharArray"]) |
+        node2.asExpr() = ma and
+        ma.getQualifier() = node1.asExpr()
+      )
+      or
+      FlowSummaryImpl::Private::Steps::summaryThroughStep(node1, node2, false)
```

Comment:
OOI: Why is this needed?

Reply:
I modelled the decode methods of shiro and Hutool in ExternalFlow.qll and ` FlowSummaryImpl::Private::Steps::summaryThroughStep(node1, node2, false)` is for using that

Reply:
As far as I know this should work by default, so you don't have to manually call it.

Reply:
I tried but it didn't work

Reply:
This doesn't work by default, because https://github.com/github/codeql/blob/4ac91ea2b97ab5a50a99e2e1df28a475c1cd7686/java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll#L251-L252 only applies to `TaintTracking::Configuration`s but not `DataFlow::Configuration`s.

I'm not sure whether changing this to a `TaintTracking::Configuration` would be the right action.

Reply:
I think it would: dataflow configs track things with the same object identity, while taint tracking also allows related but different objects. Base64 decoding clearly produces a different object but is relevant to the query. Therefore taint tracking seems appropriate (cc @aschackmull to double check you're happy with this)